### PR TITLE
Adding generic form of node perturbations

### PR DESCRIPTION
**Pull request by SabineAuer**
the idea is to make it simpler to define further perturbation scenarios. What is common to all NodePerturbations is the node number or name, the new value for the perturbed parameter, the time span of the fault and the symbol of the parameter to perturbed.
On this basis it is possible to simply define e.g. a short circuit at a node or a power perturbation. For experienced users, they can now also build the perturbations they want with "GenericPerturbation". It is possible to change any parameter of any node.

Also, I changed Line Faults to Line Failure. 

Generally, I would suggest to move all Perturbations to an extra folder.

Test are WIP.

### CODE REVIEW

**Comment by SabineAuer** [comment]
@janlisse : tests are not running through locally, yet. Also I would feel better if the Travis CI is working again.

File: src/simulations/GenericPerturbation.jl
Comment (luap-pik) [inline src/simulations/GenericPerturbation.jl:15]: Maybe rename `GenericPerturbation` -> `ChangeParameter` or sth like that? it is abit misleading that the concrete type is called "generic" while the abstract type has a more specific name :-)

**Comment by luap-pik** [comment]
btw It is a great idea to move the perturbations into a new folder.

File: src/simulations/GenericPerturbation.jl
Comment (SabineAuer) [reply src/simulations/GenericPerturbation.jl:15]: maybe I change `NodePerturbation` to `AbstractNodePerturbation` and `GenericPerturbation` to `NodeParameterChange`?

File: src/simulations/PowerPerturbation.jl
Comment (SabineAuer) [inline src/simulations/PowerPerturbation.jl:23]: @luap-pik : the tests for the power perturbation script are failing. I used a simple trick here to overcome to problem of having integers and floats for a parameter. It worked in my test example but the specific test with two SwingEqLVS nodes is running through some numeric issue.

Comment (luap-pik) [reply src/simulations/PowerPerturbation.jl:23]: I'm not sure if I understand the problem here. why do you convert to Float64, when someone passes an Int, this should be converted automatically when the rhs is called?
And in Setfield.jl you don't need to have the same type after the change.

Comment (SabineAuer) [reply src/simulations/PowerPerturbation.jl:23]: This conversion did not work automatically if you do not set a fraction of parameter change (previous implementation: multiplying initial value of parameter with fraction) but reset the absolute value of the parameter (the new way I implemented it)

Comment (luap-pik) [reply src/simulations/PowerPerturbation.jl:23]: This sounds very weird. Why does this not happen for the other perturbations?
Is the Problem that the type of the rhs changes in the fault callback?

I noticed you write `p -> p*0.0+convert(Float64, pd.power_new)`. 
shouldn't it be enough to do `p -> convert(Float64, pd.power_new)`?

Comment (luap-pik) [reply src/simulations/PowerPerturbation.jl:23]: Actually, we could do sth like `convert(eltype(getfield(np_node, np.var)), new_value)` in `mapField` to enforce type stability. This is better then hardcoding `Float64`.

Comment (SabineAuer) [reply src/simulations/PowerPerturbation.jl:23]: you are welcome to try our your suggestion and whether this fixes the tests ;)

File: src/PowerDynamics.jl
Comment (luap-pik) [inline src/PowerDynamics.jl:51]: I checked out this PR but this file seems to be missing.

Comment (SabineAuer) [reply src/PowerDynamics.jl:51]: Sry, I added some files :D that were missing

**Comment by luap-pik** [comment]
> Sry, I added some files :D that were missing 

OK, I'll have a look :-)

**Review by luap-pik — CHANGES_REQUESTED** [review]
OK I fixed the tests. Some `export` statements for `simulate`were too much. You don't need them, the general dispatch works for all perturbations.

Other errors in the tests were mainly due to inconsistencies after renaming things. Please check again carefully if this is all fine now.

Why do you have `LineFailure_new` defined? There should be a solution based on type-dispatch.

Please remove the additional entries in `Project.toml`

**Comment by SabineAuer** [comment]
@luap-pik: I generalized the AbstractNodePerturbation to AbstractPerturbation such that it can also be used for the new LineFailure fault. I moved all fault scenarios to a seperate folder. The LineFault is still working but there is a depreciation warning now when it is used. I included some more tests for better coverage. Now only the first coverage test is slightly failing. What do you think? Any more suggestions?

File: src/simulations/simulations.jl
Comment (luap-pik) [inline src/simulations/simulations.jl:6]: should the `ChangeInitialConditions` maybe also get a separate file in faults/ ?

Comment (SabineAuer) [reply src/simulations/simulations.jl:6]: We can do that. There is no simulations.jl file there anymore then.

**Comment by SabineAuer** [comment]
@luap-pik : at this point I feel ready for merging :). Even code coverage is passing now. I am proud of us ;)

**Review by luap-pik — APPROVED** [review]

**Comment by luap-pik** [comment]
@SabineAuer well done! then we're only waiting for @janlisse 's approval :-)

**Comment by SabineAuer** [comment]
@luap-pik: I think, Jan does not have the time at the moment. I will merge now. I think, we discussed this sufficiently :)

**Pull request merged**
